pkg/fleet/installer/service/file.go

```diff
@@ -0,0 +1,161 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build !windows
+
+// Package service provides a way to interact with os services
+package service
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"syscall"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// fileMutator is a struct used to transform a file
+// creating backups, replacing original files and setting permissions
+// default permissions are root:root 0644
+type fileMutator struct {
+	path             string
+	pathTmp          string
+	pathBackup       string
+	transformContent func(existing []byte) ([]byte, error)
+	validateTemp     func() error
+	validateFinal    func() error
+}
+
+// newFileMutator creates a new fileMutator
+func newFileMutator(path string, transform func(existing []byte) ([]byte, error), validateTemp, validateFinal func() error) *fileMutator {
+	return &fileMutator{
+		path:             path,
+		pathTmp:          path + ".datadog.prep",
+		pathBackup:       path + ".datadog.backup",
+		transformContent: transform,
+		validateTemp:     validateTemp,
+		validateFinal:    validateFinal,
+	}
+}
+
+func (ft *fileMutator) mutate() (rollback func() error, err error) {
+	defer os.Remove(ft.pathTmp)
+
+	originalFileExists := true
+	// create backup and temporary file if the original file exists
+	if _, err := os.Stat(ft.path); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, fmt.Errorf("could not stat file %s: %s", ft.path, err)
+		}
+		originalFileExists = false
+	}
+	if originalFileExists {
+		if err := copyFile(ft.path, ft.pathBackup); err != nil {
+			return nil, fmt.Errorf("could not create backup file %s: %s", ft.pathBackup, err)
+		}
```

Comment:
Not sure it's worth it, but on linux, we could use `renameat2` which can atomically swap files. We'd swap `path` with `pathTmp` and then rename `pathTmp` to  `pathBackup`.
I feel like it's probably safer than copying because renaming keeps all extended file attributes

Reply:
Nice but you'll still need a first copy to create the first `pathTmp` to work on.
I'll think more about the renameat2 and edge cases (when the initial file does not exist, failure in between the rename of `pathTmp` and `pathBackup`

Reply:
I prefer not to use renameat2 in the end for 2 reasons
- keep the ability to test the behaviour on darwin
- impossible to have a clean distinction between backup file and temp one: to have an atomic rollback we'll have to make temp the backup file before the swap